### PR TITLE
Add an option to apply 1-based inclusive start position format.

### PR DESCRIPTION
**Pull request by tneymanov**
We may need to contact Mayo to verify if end-pos should be inclusive. Apparently there are a bunch of formats:
https://arnaudceol.wordpress.com/2014/09/18/chromosome-coordinate-systems-0-based-1-based/

However, 1-based inclusive seems to be the most common one.

### CODE REVIEW

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 1]
```diff
@@ -265,6 +265,7 @@ def __init__(
       splittable_bgzf=False,  # type: bool
       pre_infer_headers=False,  # type: bool
       sample_name_encoding=SampleNameEncoding.WITHOUT_FILE_PATH,  # type: int
+      use_1_based_format=False,  # type: bool
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: Done.

File: gcp_variant_transforms/beam_io/vcf_parser.py [hunk 2]
```diff
@@ -544,7 +547,7 @@ def _convert_to_variant(self, record):
     self._verify_start_end(record)
     return Variant(
         reference_name=record.chrom.encode('utf-8'),
-        start=record.start,
+        start=record.pos if self._use_1_based_format else record.start,
         end=record.stop,
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: So if you check out the link in the description, it seems there are various formats of coordinate tracking. As far as I understand, 1 based inclusive is the most prominent one (Ensemble and MAF).

In 1 based inclusive, end will be inclusive, so for SNPs, start and end will have the same value. Therefore we don't modify our current end position, which is 0-based exclusive (ie, for SNP, start will be 1 less than end - USCS and Bed formats).

We probably should contact Mayo to verify that they want _inclusive_ 1 based coordinates.

File: gcp_variant_transforms/beam_io/vcfio_test.py [hunk 3]
```diff
@@ -155,6 +155,30 @@ def _get_sample_non_variant():
 
   return non_variant
 
+def _get_sample_variant_1_based():
```
Comment (samanvp) [inline gcp_variant_transforms/beam_io/vcfio_test.py:158]: The only difference here is `start=1234`, instead of duplicating the whole method could we instead add an input flag to the 3 methods (`_get_sample_variant_*`)?

File: gcp_variant_transforms/beam_io/vcf_parser.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (samanvp) [reply gcp_variant_transforms/beam_io/vcf_parser.py]: > which is 0-based exclusive
That's exactly what I wanted to know.

And yes, let's contact Mayo and make sure 1-based inclusive is what they want. We can coordinate off this thread.

File: gcp_variant_transforms/options/variant_transform_options.py [hunk 4]
```diff
@@ -100,6 +100,12 @@ def add_arguments(self, parser):
               'of the the header fields do not match the field values. Note: '
               'setting this flag or `--infer_annotation_types` incurs a '
               'performance penalty of an extra pass over all variants.'))
+    parser.add_argument(
+        '--use_1_based_coordinate',
+        type='bool', default=False, nargs='?', const=True,
+        help=('If true, start position will be 1 based, and end position will '
+              'be inclusive. Otherwise, by default the records will stored in '
+              'a 1 based, exclusive format.'))
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/options/variant_transform_options.py]: Rewrote.

File: gcp_variant_transforms/pipeline_common_test.py [hunk 5]
```diff
@@ -240,22 +240,13 @@ def test_read_variants_large_mode(self):
     assert_that(variants, asserts.count_equals_to(5))
     pipeline.run()
 
-  def test_read_variants_pysam(self):
+  def test_read_variants_use_1_based_coordinate(self):
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/pipeline_common_test.py:243]: I don't really want to test that functionality here, just the fact that nothing breaks with that flag passed. The parsing part should be verified downstream in vcfio_test (ideally actually in vcf_parser_test, but it is what it is).

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/merge_option_1_based.json [hunk 6]
```diff
@@ -0,0 +1,48 @@
+[
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/merge_option_1_based.json:1]: I don't like that.

- We could have code in run_vcf_to_bq_tests level to add num_rows and start_pos if tests use 1_based flag, but that's gonna be unnecessary complicated and not obvious in tests,
- We could have expected_result here as `1234566+1`, but it's not immediately obvious where either value comes from and why specifically here we add them but everywhere else we hardcode the value.
- We could look into creating const values for the expected results (i.e. `NUM_ROW_19=1` or something of sort), but it's a json file so not sure how we can do that. Also that would be disjointed from everything else we have and if we choose that approach (assuming it's even possible), we would need to re-haul all integration tests..

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/merge_option_1_based.json [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        "expected_result": {"sum_start": 48990}
+      },
+      {
+        "query": ["SELECT SUM(end_position) AS sum_end FROM `{DATASET_ID}.{TABLE_ID}__chr20`"],
+        "expected_result": {"sum_end": 48990}
+      },
+      {
+        "query": [
+          "SELECT COUNT(0) AS num_rows FROM `{DATASET_ID}.{TABLE_ID}__chr20` ",
+          "WHERE start_position = 14369"
```
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/merge_option_1_based.json]: Thank you, was driving me mad trying to figure out what was failing....

I thought second verification above was failing and wouldn't scroll down...

File: gcp_variant_transforms/beam_io/vcfio.py [hunk 8]
```diff
@@ -250,12 +254,15 @@ def __init__(self,
         exact data for variants and calls, without type matching.
       sample_name_encoding: specify how we want to encode sample_name mainly
         to deal with same sample_name used across multiple VCF files.
+      use_1_based_coordinate: specify whether the coordinates should be stored
+        in BQ needs on a 0 based exclusive (default) or 1 based inclusive basis.
```
Comment (samanvp) [inline gcp_variant_transforms/beam_io/vcfio.py:258]: nit: `specify whether the coordinates should be stored in BQ **using** 0 based exclusive (default) or 1 based inclusive **coordinate**.`

Also please update below.

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/merge_option_1_based.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (samanvp) [reply gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/merge_option_1_based.json:1]: Sorry, I think I was not clear what I meant: only for review to be able to compare these files side by side. So in the original json file we have:
        "expected_result": {"sum_start": 1234566}
and in the new json we will have:
        "expected_result": {"sum_start": 1234567}

So in future if someone refers back to this PR they will see the logic behind the newly added test.

I think it doesn't matter much, feel free to ignore this comment.

File: gcp_variant_transforms/vcf_to_bq.py [hunk 9]
```diff
@@ -97,7 +97,7 @@ def _read_variants(all_patterns,  # type: List[str]
                    pipeline,  # type: beam.Pipeline
                    known_args,  # type: argparse.Namespace
                    pipeline_mode,  # type: int
-                   pre_infer_headers=False  # type: bool
+                   pre_infer_headers=False,  # type: bool
```
Comment (samanvp) [inline gcp_variant_transforms/vcf_to_bq.py:100]: nit: not needed change.

File: gcp_variant_transforms/beam_io/vcfio_test.py [hunk 10]
```diff
@@ -72,7 +72,7 @@
     '19	12	.	C	<SYMBOLIC>	49	q10	AF=0.5	GT:PS:GQ	0|1:1:45	.:.:.\n')
 GVCF_LINE = '19	1234	.	C	<NON_REF>	50	.	END=1236	GT:GQ	0/0:99\n'
 
-def _get_sample_variant_1(file_name=''):
+def _get_sample_variant_1(file_name='', use_1_based_coordinate=False):
```
Comment (samanvp) [inline gcp_variant_transforms/beam_io/vcfio_test.py:75]: Please add the same logic to `_get_sample_variant_2` and 3.

Also if you could repeat `test_file_pattern_verify_details` with 1_based flags then I feel we have enough unit tests to cover this new execution branch.

File: gcp_variant_transforms/vcf_to_bq_preprocess.py [hunk 11]
```diff
@@ -107,7 +107,8 @@ def run(argv=None):
                                                all_patterns,
                                                pipeline_mode,
                                                allow_malformed_records=True,
-                                               pre_infer_headers=True)
+                                               pre_infer_headers=True,
+                                               use_1_based_coordinate=False)
```
Comment (samanvp) [inline gcp_variant_transforms/vcf_to_bq_preprocess.py:111]: Is this argument related to preprocess logic? To me, similar to `sample_name_encoding` this arg is irrelevant to preprocessor.
If that's the case let's not add this line.

File: gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/merge_option_1_based.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (tneymanov) [reply gcp_variant_transforms/testing/integration/vcf_to_bq_tests/presubmit_tests/small_tests/merge_option_1_based.json:1]: Ack.

File: gcp_variant_transforms/vcf_to_bq.py (hunk 9 above)
Comment (tneymanov) [reply gcp_variant_transforms/vcf_to_bq.py]: Done.

File: gcp_variant_transforms/vcf_to_bq_preprocess.py (hunk 11 above)
Comment (tneymanov) [reply gcp_variant_transforms/vcf_to_bq_preprocess.py]: Ahh, yeah, I had integration test failures, was trying different things. Removed.
